src/body_parser.rs

```diff
@@ -0,0 +1,124 @@
+use hyper::header::ContentType;
+use hyper::mime::{Mime, SubLevel, TopLevel};
+use serialize::{Decodable, json};
+use request::Request;
+use plugin::{Plugin, Pluggable};
+use status::StatusCode;
+use std::error::Error as StdError;
+use std::fmt;
+use std::io::{self, ErrorKind, Read};
+use typemap::Key;
+use urlencoded::{self, Query};
+
+struct BodyReader;
+
+impl Key for BodyReader {
+    type Value = String;
+}
+
+impl<'mw, 'conn, D> Plugin<Request<'mw, 'conn, D>> for BodyReader {
+    type Error = io::Error;
+
+    fn eval(req: &mut Request<D>) -> Result<String, io::Error> {
+        let mut buf = String::new();
+        try!(req.origin.read_to_string(&mut buf));
+        Ok(buf)
+    }
+}
+
+struct FormBodyParser;
+
+impl Key for FormBodyParser {
+    type Value = Query;
+}
+
+impl<'mw, 'conn, D> Plugin<Request<'mw, 'conn, D>> for FormBodyParser {
+    type Error = BodyError;
+
+    fn eval(req: &mut Request<D>) -> Result<Query, BodyError> {
+        match req.origin.headers.get::<ContentType>() {
+            Some(&ContentType(Mime(
+                TopLevel::Application,
+                SubLevel::WwwFormUrlEncoded,
+                _
+            ))) => {
+                let body = try!(req.get_ref::<BodyReader>());
+                Ok(Query::new(urlencoded::parse(&*body)))
+            },
+            _ => Err(BodyError::ContentType)
+        }
+    }
+}
+
+pub trait FormBody {
+    /// Extracts URL encoded data from the request body.
+    /// # Examples
+    /// ```{rust}
+    /// #[macro_use] extern crate nickel;
+    /// use nickel::{Nickel, HttpRouter, FormBody};
+    ///
+    /// fn main() {
+    ///     let mut server = Nickel::new();
+    ///     server.post("/a", middleware! { |req, res|
+    ///         let form_body = try_with!(res, req.form_body());
+    ///         return res.send(format!("Post: {:?}", form_body))
+    ///     });
+    /// }
+    /// ```
+    fn form_body(&mut self) -> Result<&Query, (StatusCode, BodyError)>;
+}
+
+impl<'mw, 'conn, D> FormBody for Request<'mw, 'conn, D> {
+    fn form_body(&mut self) -> Result<&Query, (StatusCode, BodyError)> {
+        self.get_ref::<FormBodyParser>().map_err(|e| (StatusCode::BadRequest, e))
+    }
+}
+
+pub trait JsonBody {
+    fn json_as<T: Decodable>(&mut self) -> Result<T, io::Error>;
+}
+
+impl<'mw, 'conn, D> JsonBody for Request<'mw, 'conn, D> {
+    fn json_as<T: Decodable>(&mut self) -> Result<T, io::Error> {
```

Comment:
Can you add a FIXME here to update the error type on `json_as` when we rename/break it in future? Would be good to capture that possible Parsing error rather than a generic `io::Error`.